src/test/ui/consts/const-address-of.rs

```diff
@@ -0,0 +1,16 @@
+// check-pass
+
+#![feature(raw_ref_op)]
+
+const A: *const i32 = &raw const *&2;
+static B: () = { &raw const *&2; };
+static mut C: *const i32 = &raw const *&2;
```

Comment:
Please also add tests for `&raw const 2` and `&raw mut 2`, I don't think I saw any.

Reply:
`&raw [const|mut] 2` isn't allowed since `2` isn't a place expression. I've added some tests that assign to a variable first though.

Reply:
Oh, there are no tests in this PR because the tests already exist but didn't change in this PR?

Reply:
Hmm.. I can't find anything. Do we have any tests showing the error happening when one tries to do this even though it's not allowed?

Reply:
https://github.com/rust-lang/rust/blob/a605441e049f0b6d5f7715b94b8ac4662fd7fcf6/src/test/ui/raw-ref-op/raw-ref-temp.rs#L9

Reply:
Ah, so this is checked on the HIR, and already mentions the downside to that:

https://github.com/rust-lang/rust/blob/8843b28e64b02c2245f5869ad90cafa6d85ab0d9/src/librustc_typeck/check/expr.rs#L494-L496

You could implement this check on the MIR by bailing out if any `Rvalue::AddressOf`'s `Place` is a temporary. Although that will probably start failing `&raw const *&*local` (reborrowing a local and then reading from it). Can we somehow reuse the logic that causes 

```rust
fn foo() -> String {
    unimplemented!()
}

fn main() {
    let x = foo().as_str();
    println!("{}", x);
}
```

to emit

```
error[E0716]: temporary value dropped while borrowed
 --> src/main.rs:6:13
  |
6 |     let x = foo().as_str();
  |             ^^^^^         - temporary value is freed at the end of this statement
  |             |
  |             creates a temporary which is freed while still in use
7 |     println!("{}", x);
  |                    - borrow later used here
  |
  = note: consider using a `let` binding to create a longer lived value
```

This is not a blocker for this PR, but just for stabilization imo.

So r=me with CI passing (you need to fix a typo in a test)

Reply:
I think the right place for that design discussion is over at https://github.com/rust-lang/rust/issues/66708.